src/sentry/spans/consumers/process/factory.py

```diff
@@ -119,28 +156,39 @@ def _produce_segment(message: Message[ProduceSegmentContext | None]):
             if len(keys) > 0:
                 payload_context["sample_key"] = keys[0]
 
-            total_spans_read = 0
             # With pipelining, redis server is forced to queue replies using
             # up memory, so batching the keys we fetch.
             with txn.start_child(op="process", description="produce_fetched_segments"):
                 for i in range(0, len(keys), BATCH_SIZE):
                     segments = client.read_and_expire_many_segments(keys[i : i + BATCH_SIZE])
 
                     for segment in segments:
-                        total_spans_read += len(segment)
-                        produce_segment_to_kafka(segment)
+                        if not segment:
+                            continue
 
-            sentry_sdk.set_context("payload", payload_context)
+                        payload_data = prepare_buffered_segment_payload(segment)
+                        buffered_segments.append(KafkaPayload(None, payload_data, []))
 
+    return buffered_segments
 
-def produce_segment(message: Message[ProduceSegmentContext | None]):
+
+def expand_segments(context_dict: dict[int, ProduceSegmentContext]):
     try:
-        _produce_segment(message)
+        return _expand_segments(context_dict)
     except Exception:
         sentry_sdk.capture_exception()
+        return []
 
 
 class ProcessSpansStrategyFactory(ProcessingStrategyFactory[KafkaPayload]):
+    """
+    1. Process spans and push them to redis
+    2. Commit offsets for processed spans
+    3. Reduce the messages to find timestamps we need to process segments for
+    4. Fetch all segments that are ready to be processed
+    5. Produce segments to buffered-segments topic
```

Comment:
is it worth mentioning here that the segment is removed from redis? otherwise 3-5 may be repeated for the same segment